File: typescript/src/hooks/stableSurgeHook.ts
import { HookBase, HookStateBase } from './types';
import { MathSol } from '../utils/math';
import { SwapKind, SwapParams } from '../vault/types';
import { Stable } from '../stable';

export type HookStateStableSurge = HookStateBase & {
    hookType: 'StableSurge';
    amp: bigint;
    surgeThresholdPercentage: bigint;
    maxSurgeFeePercentage: bigint;
};

// Implementation from mono-repo commit: 1c9d6a2913eb2d1210019455b44192760d9beb03
export class StableSurgeHook implements HookBase {
    public shouldCallComputeDynamicSwapFee = true;
    public shouldCallBeforeSwap = false;
    public shouldCallAfterSwap = false;
    public shouldCallBeforeAddLiquidity = false;
    public shouldCallAfterAddLiquidity = false;
    public shouldCallBeforeRemoveLiquidity = false;
    public shouldCallAfterRemoveLiquidity = false;
    public enableHookAdjustedAmounts = false;

    onComputeDynamicSwapFee(
        params: SwapParams,
        pool: string,
        staticSwapFeePercentage: bigint,
        hookState: HookStateStableSurge,
    ): { success: boolean; dynamicSwapFee: bigint } {
        const stablePool = new Stable(hookState);

        return {
            success: true,
            dynamicSwapFee: this.getSurgeFeePercentage(
                params,
                stablePool,
                hookState.surgeThresholdPercentage,
                hookState.maxSurgeFeePercentage,
                staticSwapFeePercentage,
            ),
        };
    }

    private getSurgeFeePercentage(
        params: SwapParams,
        pool: Stable,
        surgeThresholdPercentage: bigint,
        maxSurgeFeePercentage: bigint,
        staticFeePercentage: bigint,
    ): bigint {
        const amountCalculatedScaled18 = pool.onSwap(params);
        const newBalances = [...params.balancesLiveScaled18];

        if (params.swapKind === SwapKind.GivenIn) {
            newBalances[params.indexIn] += params.amountGivenScaled18;
            newBalances[params.indexOut] -= amountCalculatedScaled18;
        } else {
            newBalances[params.indexIn] += amountCalculatedScaled18;
            newBalances[params.indexOut] -= params.amountGivenScaled18;
        }

        const newTotalImbalance = this.calculateImbalance([...newBalances]);

        // If we are balanced, or the balance has improved, do not surge: simply return the regular fee percentage.
        if (newTotalImbalance === 0n) {
            return staticFeePercentage;
        }

        const oldTotalImbalance = this.calculateImbalance([
            ...params.balancesLiveScaled18,
        ]);

        if (
            newTotalImbalance <= oldTotalImbalance ||
            newTotalImbalance <= surgeThresholdPercentage
        ) {
            return staticFeePercentage;
        }

        // surgeFee = staticFee + (maxFee - staticFee) * (pctImbalance - pctThreshold) / (1 - pctThreshold).
        //
        // As you can see from the formula, if it’s unbalanced exactly at the threshold, the last term is 0,
        // and the fee is just: static + 0 = static fee.
        // As the unbalanced proportion term approaches 1, the fee surge approaches: static + max - static ~= max fee.
        // This formula linearly increases the fee from 0 at the threshold up to the maximum fee.
        // At 35%, the fee would be 1% + (0.95 - 0.01) * ((0.35 - 0.3)/(0.95-0.3)) = 1% + 0.94 * 0.0769 ~ 8.2%.
        // At 50% unbalanced, the fee would be 44%. At 99% unbalanced, the fee would be ~94%, very close to the maximum.
        const dynamicSwapFee =
            staticFeePercentage +
            MathSol.mulDownFixed(
                maxSurgeFeePercentage - staticFeePercentage,
                MathSol.divDownFixed(
                    newTotalImbalance - surgeThresholdPercentage,
                    MathSol.complementFixed(surgeThresholdPercentage),
                ),
            );
        return dynamicSwapFee;
    }

    private calculateImbalance(balances: bigint[]): bigint {
        const median = this.findMedian(balances);

        let totalBalance = 0n;
        let totalDiff = 0n;

        for (let i = 0; i < balances.length; i++) {
            totalBalance += balances[i];
            totalDiff += this.absSub(balances[i], median);
        }

        return MathSol.divDownFixed(totalDiff, totalBalance);
    }

    private findMedian(balances: bigint[]): bigint {
        const sortedBalances = balances.sort((a, b) => {
            if (a < b) return -1;
            if (a > b) return 1;
            return 0;
        });
        const mid = Math.floor(sortedBalances.length / 2);

        if (sortedBalances.length % 2 == 0) {
            return (sortedBalances[mid - 1] + sortedBalances[mid]) / 2n;
        } else {
            return sortedBalances[mid];
        }
    }

    private absSub(a: bigint, b: bigint): bigint {
        return a > b ? a - b : b - a;
    }

    onBeforeAddLiquidity() {
        return { success: false, hookAdjustedBalancesScaled18: [] };
    }

    onAfterAddLiquidity() {
        return { success: false, hookAdjustedAmountsInRaw: [] };
    }

    onBeforeRemoveLiquidity() {
        return { success: false, hookAdjustedBalancesScaled18: [] };
    }

    onAfterRemoveLiquidity() {
        return { success: false, hookAdjustedAmountsOutRaw: [] };
    }

    onBeforeSwap() {
        return { success: false, hookAdjustedBalancesScaled18: [] };
    }

    onAfterSwap() {
        return { success: false, hookAdjustedAmountCalculatedRaw: 0n };
    }
}

export default StableSurgeHook;

File: python/src/hooks/stable_surge_hook.py
from typing import Dict, List

from src.maths import (
    div_down_fixed,
    mul_down_fixed,
    complement_fixed,
)
from src.pools.stable import Stable
from src.swap import SwapKind


# This hook implements the StableSurgeHook found in mono-repo: https://github.com/balancer/balancer-v3-monorepo/blob/main/pkg/pool-hooks/contracts/StableSurgeHook.sol
class StableSurgeHook:
    def __init__(self):
        self.should_call_compute_dynamic_swap_fee = True
        self.should_call_before_swap = False
        self.should_call_after_swap = False
        self.should_call_before_add_liquidity = False
        self.should_call_after_add_liquidity = False
        self.should_call_before_remove_liquidity = False
        self.should_call_after_remove_liquidity = False
        self.enable_hook_adjusted_amounts = False

    def on_before_add_liquidity(self):
        return {"success": False, "hook_adjusted_balances_scaled18": []}

    def on_after_add_liquidity(self):
        return {"success": False, "hook_adjusted_amounts_in_raw": []}

    def on_before_remove_liquidity(self):
        return {"success": False, "hook_adjusted_balances_scaled18": []}

    def on_after_remove_liquidity(self):
        return {"success": False, "hook_adjusted_amounts_out_raw": []}

    def on_before_swap(self):
        return {"success": False, "hook_adjusted_balances_scaled18": []}

    def on_after_swap(self):
        return {"success": False, "hook_adjusted_amount_calculated_raw": 0}

    def on_compute_dynamic_swap_fee(
        self,
        params: Dict,
        static_swap_fee_percentage: int,
        hook_state: Dict,
    ) -> Dict[str, int]:
        stable_pool = Stable(hook_state)

        return {
            "success": True,
            "dynamic_swap_fee": self.get_surge_fee_percentage(
                params,
                stable_pool,
                hook_state["surgeThresholdPercentage"],
                hook_state["maxSurgeFeePercentage"],
                static_swap_fee_percentage,
            ),
        }

    def get_surge_fee_percentage(
        self,
        params: Dict,
        pool: Stable,
        surge_threshold_percentage: int,
        max_surge_fee_percentage: int,
        static_fee_percentage: int,
    ) -> int:
        amount_calculated_scaled_18 = pool.on_swap(params)
        new_balances = params["balances_live_scaled18"][:]

        if params["swap_kind"] == SwapKind.GIVENIN.value:
            new_balances[params["index_in"]] += params["amount_given_scaled18"]
            new_balances[params["index_out"]] -= amount_calculated_scaled_18
        else:
            new_balances[params["index_in"]] += amount_calculated_scaled_18
            new_balances[params["index_out"]] -= params["amount_given_scaled18"]

        new_total_imbalance = self.calculate_imbalance(new_balances)

        if new_total_imbalance == 0:
            return static_fee_percentage

        old_total_imbalance = self.calculate_imbalance(params["balances_live_scaled18"])

        if (
            new_total_imbalance <= old_total_imbalance
            or new_total_imbalance <= surge_threshold_percentage
        ):
            return static_fee_percentage

        dynamic_swap_fee = static_fee_percentage + mul_down_fixed(
            max_surge_fee_percentage - static_fee_percentage,
            div_down_fixed(
                new_total_imbalance - surge_threshold_percentage,
                complement_fixed(surge_threshold_percentage),
            ),
        )
        return dynamic_swap_fee

    def calculate_imbalance(self, balances: List[int]) -> int:
        median = self.find_median(balances)

        total_balance = sum(balances)
        total_diff = sum(self.abs_sub(balance, median) for balance in balances)

        return div_down_fixed(total_diff, total_balance)

    def find_median(self, balances: List[int]) -> int:
        sorted_balances = sorted(balances)
        mid = len(sorted_balances) // 2

        if len(sorted_balances) % 2 == 0:
            return (sorted_balances[mid - 1] + sorted_balances[mid]) // 2
        else:
            return sorted_balances[mid]

    def abs_sub(self, a: int, b: int) -> int:
        return abs(a - b)
